### PR TITLE
improve `estimate_gas_with`

### PR DESCRIPTION
**Pull request by tcoratger**
- At the beginning of the function `highest_gas_limit` is truly defined as the maximum between the block env gas limit and the transaction request gas limit (if both exists, then a maximum is computed in the mapping).
- A few variables are renamed for clarity:
   - `tx_request_gas_limit` being the gas limit for the transaction request.
   - `tx_request_gas_price` being the gas price for the transaction request.
   - `block_env_gas_limit` being the gas limit of the block environment.
 - Some complementary explanations are given inside the documentation for some tricky stuffs which weren't super clear to me on first reading.
 - The `highest_gas_limit` is converted to `u64` earlier to avoid confusion with U256 as this is not needed here.

But the logic still is the same.

### CODE REVIEW

File: crates/rpc/rpc/src/eth/api/call.rs [hunk 1]
```diff
@@ -320,15 +323,16 @@ where
             )?;
         };
 
-        // pick a point that's close to the estimated gas
+        // Pick a point that's close to the estimated gas
         let mut mid_gas_limit = std::cmp::min(
             gas_used * 3,
```
Comment (tcoratger) [inline crates/rpc/rpc/src/eth/api/call.rs:328]: Here as we have potentially done in the meantime a new simulation with the `optimistic_gas_limit` we should not rather use `res.result.gas_used()` which can be different from `gas_used` in the case where the condition `optimistic_gas_limit < highest_gas_limit` was verified previously.

**Review by mattsse — CHANGES_REQUESTED** [review]
thanks!

pedantic style nit

File: crates/storage/storage-api/src/block_id.rs
[the remark this answers is not in the document]
Comment (tcoratger) [reply crates/storage/storage-api/src/block_id.rs:1]: My bad, I did this change and then forgot, will take care next time.

File: crates/rpc/rpc/src/eth/api/call.rs [hunk 2]
```diff
@@ -198,31 +198,35 @@ where
...
             apply_state_overrides(state_override, &mut db)?;
         }
-        // if the request is a simple transfer we can optimize
+
+        // Optimize for simple transfer transactions, potentially reducing the gas estimate.
         if env.tx.data.is_empty() {
             if let TransactTo::Call(to) = env.tx.transact_to {
                 if let Ok(code) = db.db.account_code(to) {
-                    let no_code_callee = code.map(|code| code.is_empty()).unwrap_or(true);
+                    let no_code_callee = code.map_or(true, |code| code.is_empty());
```
[the remark this answers is not in the document]
Comment (tcoratger) [reply crates/rpc/rpc/src/eth/api/call.rs]: Fixed

File: crates/rpc/rpc/src/eth/api/call.rs [hunk 3]
```diff
@@ -198,31 +198,35 @@ where
...
+        let tx_request_gas_limit = request.gas;
+        let tx_request_gas_price = request.gas_price;
+        let block_env_gas_limit = block.gas_limit;
+
+        // Determine the highest possible gas limit, considering both the request's specified limit
+        // and the block's limit.
+        let mut highest_gas_limit = tx_request_gas_limit
+            .map_or(block_env_gas_limit, |tx_gas_limit| {
+                U256::from(tx_gas_limit).max(block_env_gas_limit)
+            });
```
[the remark this answers is not in the document]
Comment (tcoratger) [reply crates/rpc/rpc/src/eth/api/call.rs]: fixed

File: crates/rpc/rpc/src/eth/api/call.rs (hunk 1 above)
Comment (tcoratger) [reply crates/rpc/rpc/src/eth/api/call.rs:327]: @mattsse What do you think about this?

**Review by mattsse — APPROVED** [review]

File: crates/rpc/rpc/src/eth/api/call.rs (hunk 1 above)
Comment (mattsse) [reply crates/rpc/rpc/src/eth/api/call.rs:327]: not following tbh

Comment (tcoratger) [reply crates/rpc/rpc/src/eth/api/call.rs:327]: I just wanted to say that here the `gas_used` variable used for the calculation is still the old variable defined above with `let gas_used = res.result.gas_used();`. But between this declaration and the use of `gas_used` here, in the case where `if optimistic_gas_limit < highest_gas_limit` is true, a new execution is done with `(res, env) = self.transact(&mut db, env)?;` . As in this case `res` is updated, `res.result.gas_used()` may be different from its old value before this re-execution.

So I was wondering if on this line it wasn't better to put `res.result.gas_used() * 3` directly, wdyt?

Comment (tcoratger) [reply crates/rpc/rpc/src/eth/api/call.rs:327]: Fixed https://github.com/paradigmxyz/reth/pull/8541

**Pull request merged**